rest-java/src/test/java/org/hiero/mirror/restjava/controller/NetworkControllerTest.java

```diff
@@ -524,4 +529,299 @@ void notFound() {
+            "?timestamp=gt:{0}&timestamp=lt:{2}, 1"
+        })
+        void timestampBounds(String parameters, int expectedIndex) {
+            // given - create account balances at different timestamps
+            final var timestamp1 = domainBuilder.timestamp();
+            final var timestamp2 = domainBuilder.timestamp();
+            createUnreleasedSupplyAccounts(timestamp1);
+            createUnreleasedSupplyAccounts(timestamp2);
+            final var timestamp0 = timestamp1 - 1000000;
+
+            final var formattedParams = MessageFormat.format(
+                    parameters,
+                    commonMapper.mapTimestamp(timestamp0),
+                    commonMapper.mapTimestamp(timestamp1),
+                    commonMapper.mapTimestamp(timestamp2));
+
+            final var expected0Timestamp = commonMapper.mapTimestamp(timestamp0);
+            final var expected1Timestamp = commonMapper.mapTimestamp(timestamp1);
+            final var expected2Timestamp = commonMapper.mapTimestamp(timestamp2);
+            final var expectedTimestamps = List.of(expected0Timestamp, expected1Timestamp, expected2Timestamp);
+
+            // when
+            final var response =
+                    restClient.get().uri(formattedParams).retrieve().body(NetworkSupplyResponse.class);
+
+            // then
+            assertThat(response).isNotNull();
+            assertThat(response.getTimestamp()).isEqualTo(expectedTimestamps.get(expectedIndex));
+        }
+
+        @Test
+        void timestampDeduplicated() {
+            // given - multiple balances at different timestamps, ensure latest is used
+            final var timestamp1 = 1_600_000_000_000_000_000L;
+            final var timestamp2 = 1_700_000_000_000_000_000L;
+            final var timestamp3 = 1_700_000_000_000_000_005L;
+
+            // Create different balances at different timestamps for same accounts
+            createCustomBalance(2, 10L, timestamp1);
+            createCustomBalance(42, 20L, timestamp1);
+            createCustomBalance(2, 1L, timestamp2);
+            createCustomBalance(42, 1L, timestamp2);
+            createCustomBalance(2, 4_000_000_000_000_000_000L, timestamp3);
+            createCustomBalance(42, 50L, timestamp3);
+
+            // when - query for timestamp range that includes latest data
+            final var response = restClient
+                    .get()
+                    .uri("?timestamp=lte:1700000000.000000000")
+                    .retrieve()
+                    .body(NetworkSupplyResponse.class);
+
+            // then - should use latest balances within range (timestamp2)
+            assertThat(response).isNotNull();
+            assertThat(response.getTimestamp()).isEqualTo("1700000000.000000000");
+        }
+
+        private void createCustomBalance(long accountNum, long balance, long timestamp) {
+            final var accountId = EntityId.of(0, 0, accountNum);
```

Comment:
This is going to fail tests in the Non-Zero Realm workflow. This suggestion should do the work:
```suggestion
            final var entityIdParameter = EntityIdParameter.valueOf(String.valueOf(accountNum));
            final var accountId = EntityId.of(entityIdParameter.shard(), entityIdParameter.realm(), accountNum);
```